### ISSUE TITLE
Support specifying the username in state locks

### ISSUE THREAD

**Issue by ag-TJNII** [opentofu/opentofu#2007]
### OpenTofu Version

```shell
OpenTofu v1.8.2
on linux_amd64
```

### Use Cases

Currently Tofu uses the system `[username]@[hostname]` for the `Who` field in the lockfile.  This is fine if TF is running as a uniquely named user on a uniquely named host.  However, this can fall apart when TF is run in a wrapper.

We distribute TF to users as internal Docker containers, which has been a boon to maintainability and overall has suited us very well.  However, for locking this results in the lock always showing `tfuser@[random ID]` as the user name in the container is fixed and the container hostname is random.

I would like to be able to set this as a backend setting to allow setting the `who` field to something a operator will recognize, to allow the reported error message to be properly actionable. 

### Attempted Solutions

- Confirmed this appears to be a hardcode: https://github.com/opentofu/opentofu/blob/52eb928e6ad4daeacf70a6bb62f321897bd846ce/internal/states/statemgr/locker.go#L169

### Proposal

Allow the Who field in the LockInfo object to be specified as config for automated environments where `[username]@[hostname]` is not sufficiently unique.

**Comment by haiderbari** [comment]
Hey there, you need to give a username with the http remote backend configuration. Not sure which other backend configs that needs username set. I always wanted who to be set to the username passed into http backend config.

**Comment by haiderbari** [comment]
Hi core team any update on this feature. I think it makes more sense to the use the username from the backend config to set the WHO field instead of using the host and username from the home directory path. If any backend that needs to specify a username in it's config then it should really use that to set the WHO field.

**Comment by yottta** [comment]
Hi @haiderbari, thanks for the follow-up.
Just as a personal curiosity, would something like this help with your situation? https://github.com/opentofu/opentofu/issues/2611
The `who` field will remain untouched but maybe some additional information that could help identifying the actual user.

**Comment by haiderbari** [comment]
Hi @yottta not really because as it dependent on the user to set it. Here my use case since I am using a HTTP backend the information returned back to the user if the state is locked is important because it has the data when the lock was created and who locked it. As ag-TJNII has mentioned when executing tofu commands within a docker image or on a EC2 instance its not helpful on who locked as hard to recognize the user. Since the http backend config needs the username set why not use it or any other backend config that uses it as well. Else use the host and system name that don't use it like S3

**Comment by yottta** [comment]
Discussed today with the core team, and touching the existing `who` field is not something that we plan to do.
Instead, the suggestion from #2611 is something that we'd probably do eventually.
Though, this is low priority right now since there are some plans on reworking the current implementation of the backends, and we need to have some more clarity on how we are going to do it.

**Comment by cam72cam** [comment]
We are considering backends as plugins #382, which would allow much easier customization of how different backends function.  Adding the ability to provide misc info in the lock metadata would likely be included.

**Comment by acolombier** [comment]
Hi, I understand the backends as plugins would solve this, tho as I read, it seems this would be a significant piece of effort to implement. As this issue as been open for more than a year, and the backend-as-plugin for more than two, would you accept a minor change to allow customising the current info, according to the suggested plan in [#3641](https://github.com/opentofu/opentofu/issues/3641)?

**Comment by apparentlymart** [comment]
Almost everything related to the functionality of backends is on hold until we better understand the path for backends as plugins, because our preferred path is to follow the state storage plugin protocol designed by HashiCorp for Terraform so that those who want to implement these plugins won't need to redundantly implement them twice to support both products.

This issue unfortunately comes under that hold because if we were to implement a special solution for it in OpenTofu that is incompatible with the plugin protocol then we would not be able to adopt the plugin protocol without breaking that functionality.

We will revisit this again once we have a final understanding of the protocol that Terraform is using and have decided whether it'd be suitable for OpenTofu's needs or whether we're going to design our own OpenTofu-specific protocol. At that point we'll make the final call about whether this is something we can implement as a general feature across all backends, or whether it's something that each backend would need to implement itself.

It looks like we missed this before when labeling about that hold, so I'm going to retroactively add labels to represent that this is blocked and that it's a "provider protocol" concern. (We're assuming for the moment that the state storage protocol will be part of the provider protocol, but even if not it'll require _some_ protocol to be designed and implemented.)

---

In the meantime I suggest designing your systems on the assumption that OpenTofu itself does not track any useful metadata about locks at all, and so arrange to track the information you need in some other way.

What "other ways" are possible will of course vary depending on what backend you've chosen to use, but e.g. in Amazon S3 you can use the `lock_tags` argument to tell the state storage implementation to add extra [tags](https://docs.aws.amazon.com/AmazonS3/latest/userguide/object-tagging.html) to the lock objects, setting one of the tags to "something an operator will recognize".

**Comment by acolombier** [comment]
Thanks for you feedback.

> In the meantime I suggest designing your systems on the assumption that OpenTofu itself does not track any useful metadata about locks at all, and so arrange to track the information you need in some other way.

I am not sure this is neither correct and helpful. 
As we can see [here](https://github.com/opentofu/opentofu/blob/56b63795afa338196d3007471728420c5bce135b/internal/states/statemgr/locker.go#L136), OpenTofu (or Terraform legacy) already tracks this information, but as I explained in #3641, this is often flawed when operated from ephemeral environment (CI or (dev)container), which is very common nowadays (if not the most common usecase when applying infra changes).

I understand we don't want to increase the number of information tracked by the lock, but in this case, this is information that is already being tracked.

> you can use the `lock_tags` argument

There is a few problem with this:

1. OpenTofu wouldn't consume this data. So for example, in the case of CI, those would not display this useful information in the timely manner, and would either have convoluted logic (e.g some awscli command to run after if it "fails" - quotes to highlight that failure doesn't always mean the state lock, but also that concurrency element would be lost, meaning that by the time that command runs, the lock could have been removed) or impossible (an automated nightly CI which would have fail during the night couldn't be diagnose in the following morning - except of course if the lock object is still there)
2. The data is optional, meaning that it could be accidentally omitted, potentially invalidating any workaround for the point `1)`

Perhaps I should re-open #3641 as a bug instead? Curious to hear @diofeher's opinion.

**Comment by apparentlymart** [comment]
To be clear, I was intending to say that the information OpenTofu currently tracks _is not useful_ -- it's capturing irrelevant implementation details about the environment where OpenTofu happens to be running -- not that OpenTofu does not track any information.

I don't think reopening https://github.com/opentofu/opentofu/issues/3641 would significantly change the situation. The overall problem is that the model of state storage is about to change in a very significant way -- moving most of the responsibility into plugins instead of OpenTofu itself -- and so any work we did in this area today has a high likelihood of being made immediately irrelevant as we adopt the new architecture. That would then potentially leave us in the position of having to choose between either breaking functionality we recently released or making it considerably harder to write state storage plugins that can work with both OpenTofu and Terraform.

In case I wasn't clear about this in my earlier comment: we _do_ want to support a model where operators can configure arbitrary metadata to be saved with a state lock and then retrieved later to display as part of the error message about lock contention. The only question is exactly how that might fit into the new plugin-based state storage model, which we currently cannot predict because the protocol for that is not yet finalized. We will consider this again once the protocol is finalized.

(The adoption of plugin-based state storage might cause even the useless information OpenTofu _currently_ captures to no longer be possible, which I expect we'd consider acceptable _because_ it is essentially useless. But the same would not be true for a feature intentionally designed to allow you to capture useful information where we'd be committing to support it indefinitely.)

**Comment by acolombier** [comment]
> any work we did in this area today has a high likelihood of being made immediately irrelevant as we adopt the new architecture

As I understand from #382, this change has been awaited for more than 2 years. I see there was some update three month ago to get the prototype ready (#3476) but this still looks like some effort will have to happen before this is actually replacing the current state lock info, such as implementing the actual plugins.

So the question is, wouldn't it be worth getting this low hanging fruit change and remove it, as the state lock gets entirely dropped and ? I have already implemented them locally, just would need to add tests if we were going ahead with it. 

Best case scenario, these few lines changes get yanked in case the plugins indeed get ready to go and replace the state lock.

Worse case scenario, there is a workaround for this problem, which can safely be superseded when the plugins takes over, whatever time it takes.

**Comment by apparentlymart** [comment]
Our compatibility promises mean that any feature we include in a stable release must typically remain working forever.

We do have some exceptions for cases like functionality becoming truly obsolete -- for example, support for operating systems that are no longer supported by their own vendors -- but it doesn't seem like _this_ functionality would fit any of that criteria.

I understand that it's frustrating to wait. We are also quite frustrated at how long it's taking for this question to be answered. But I'm afraid the situation is what it is: we will, as I previously stated, revisit this only once it's clear how plugin-based state storage is going to work, and that in turn depends on how the Terraform protocol for it is designed to work.

I'm glad to hear that you've been able to make a small fork of OpenTofu with an interim solution that works for you. I hope you will be able to continue using that modified version in the meantime.
